app/server/appsmith-plugins/mongoPlugin/src/main/java/com/external/plugins/MongoPlugin.java

```diff
@@ -232,8 +232,14 @@ private String getDatabaseName(DatasourceConfiguration datasourceConfiguration)
              */
 
             return Mono.just(datasourceConfiguration)
-                    .map(MongoPluginExecutor::buildClientURI)
-                    .map(MongoClients::create)
+                    .map(dsConfig -> {
+                        try {
+                            return buildClientURI(dsConfig);
+                        } catch (AppsmithPluginException e) {
+                            return e;
+                        }
+                    })
+                    .map(uri -> MongoClients.create((String)uri))
```

Comment:
We should almost never need to typecast the values we get inside `map` and `flatMap`. If we are needing to do it, most of the time (not always), we're missing something. (Lesson learnt the hard way 😉).

Here, the function to the first `.map` call doesn't *always* return a `String`. It's returning an exception object when there's an error. When this happens, we'd get a class cast exception on hour hands when we try to cast it to a `String`. Here's a better way:

```suggestion
                    .flatMap(dsConfig -> {
                        try {
                            return Mono.just(buildClientURI(dsConfig));
                        } catch (AppsmithPluginException e) {
                            return Mono.error(e);
                        }
                    })
                    .map(MongoClients::create)
```

Reply:
Thanks @sharat87